pkg/waiter/manager.go

```diff
@@ -0,0 +1,194 @@
+	}
+}
+
+// Start spawns a goroutine that will run forever or until ctx done delivering
+// messages to waiters.
+func (w *managerImpl[T]) Start(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				// ensure are no more sends.
+				w.managerShutdownSignal.Signal()
+
+				// inform the waiters.
+				w.closeWaiters()
+				return
+			case r := <-w.responseCh:
+				waiterCh, found := w.removeWaiter(r.id)
+				if !found {
+					continue
+				}
+
+				// to prevent blocking, waiterCh should be buffered with
+				// size of 1 (each waiterCh should only ever receive 1 msg)
+				// (alternatively use a goroutine).
+				waiterCh <- r
+			case id := <-w.doneWaiterCh:
+				// the waiter has been closed or canceled.
+				w.removeWaiter(id)
+			}
+		}
+	}()
+}
+
+// Send sends data and err to the waiter with the provided id.
+func (w *managerImpl[T]) Send(id string, data T, err error) error {
+	if w.managerShutdownSignal.IsDone() {
+		return ErrManagerShutdown
+	}
+
+	// check again if the manager is shutdown, return if so
+	// otherwise write the message to the proper channel.
+	select {
+	case <-w.managerShutdownSignal.Done():
+		return ErrManagerShutdown
+	case w.responseCh <- &response[T]{
+		id:   id,
+		data: data,
+		err:  err,
+	}:
+	}
+
+	return nil
+}
+
+// NewWaiter creates a waiter with a unique ID. A call to Wait() will complete when
+// a response is published using this ID.
+func (w *managerImpl[T]) NewWaiter() (Waiter[T], error) {
+	// if the manager is shutdown, error out.
+	if w.managerShutdownSignal.IsDone() {
+		return nil, ErrManagerShutdown
+	}
+
+	// otherwise setup the new waiter.
+	for i := 0; i < maxIDCollisions; i++ {
+		id := uuid.NewV4().String()
+
+		waiterCh, created := w.addWaiter(id)
+		if created {
+			return newGenericWaiter(id, waiterCh, w.doneWaiterCh), nil
+		}
+	}
+
+	return nil, ErrTooManyCollisions
+}
+
+// addWaiter will return true if the waiter chan was successful created, otherwise
+// will return false indicating an id collision.
+func (w *managerImpl[T]) addWaiter(id string) (chan *response[T], bool) {
+	w.waitersMu.Lock()
+	defer w.waitersMu.Unlock()
+
+	if _, ok := w.waiters[id]; ok {
+		return nil, false
+	}
+
+	// create a buffered channel of size 1 so that the loop initiated by w.Start()
+	// will not have to wait for this channel to be read (making it non-blocking).
+	ch := make(chan *response[T], 1)
+
+	w.waiters[id] = ch
+	return ch, true
+}
+
+// removeWaiter will return true if a waiter was found and removed with the waiters
+// chan, otherwise will return false indicating no waiter found.
+func (w *managerImpl[T]) removeWaiter(id string) (chan *response[T], bool) {
+	w.waitersMu.Lock()
+	defer w.waitersMu.Unlock()
+
+	waiterCh, ok := w.waiters[id]
+	if !ok {
+		return nil, false
+	}
+
+	delete(w.waiters, id)
+
+	return waiterCh, true
+}
+
+func (w *managerImpl[T]) closeWaiters() {
+	w.waitersMu.Lock()
+	defer w.waitersMu.Unlock()
+
+	for _, ch := range w.waiters {
```

Comment:
might be good to delete each entry from the map and then set the map to `nil` to help the GC know it can clean this all up